pkg/util/push/push.go

```diff
@@ -17,14 +17,16 @@ func Handler(cfg distributor.Config, push func(context.Context, *client.WriteReq
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		compressionType := util.CompressionTypeFor(r.Header.Get("X-Prometheus-Remote-Write-Version"))
 		var req client.PreallocWriteRequest
-		req.Source = client.API
 		_, err := util.ParseProtoReader(r.Context(), r.Body, int(r.ContentLength), cfg.MaxRecvMsgSize, &req, compressionType)
 		logger := util.WithContext(r.Context(), util.Logger)
 		if err != nil {
 			level.Error(logger).Log("err", err.Error())
 			http.Error(w, err.Error(), http.StatusBadRequest)
 			return
 		}
+		if req.Source == 0 {
+			req.Source = client.API
+		}
```

Comment:
Getting back to this change, @bboreham as right and I was wrong. The `req.Source` is not overridden by the unmarshalling if not included in the request, so we should have fixed the issue to the root: ensuring `Source` is always set within Cortex.

https://github.com/cortexproject/cortex/blob/master/pkg/ingester/client/cortex.pb.go#L5502-L5520

@Wing924 May you re-iterate on this?

Reply:
> The req.Source is not overridden by the unmarshalling if not included in the request
It's not true.

Let's change the code like this:
```diff
+ req.Source = 2
		_, err := util.ParseProtoReader(r.Context(), r.Body, int(r.ContentLength), cfg.MaxRecvMsgSize, &req, compressionType)
		logger := util.WithContext(r.Context(), util.Logger)
		if err != nil {
			level.Error(logger).Log("err", err.Error())
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
- if req.Source == 0 {
- 	req.Source = client.API
- }
```
if `Source` is not included in the request, we expect `req.Source` is `2`, right?
But actually it's `0`.
You can check my test.

Reply:
https://github.com/cortexproject/cortex/blob/master/pkg/util/http.go#L107
this line reset req.Source to `0`

Reply:
How about you explain your change, where I have substituted one const value with its definition for clarity:
```
if req.Source == 0 {
	req.Source = 0
}
```

Reply:
I know `client.API` is equals to `0` so that you think these code is useless.
This code is explicitly set default value if req.Source is not set.

Reply:
Physically, `client.API` equals to `0`, but logically, they are different things.

Reply:
OK, the `Reset()` is indeed the important line, thanks.
So the original code didn't do anything (it set to 0 which was then overridden to 0) and the new code doesn't do anything.